lib/src/annotate.rs

```diff
@@ -0,0 +1,397 @@
+        new_commit_id: &CommitId,
+        new_local_line_number: usize,
+    ) {
+        let old_map = self.local_line_map.get_mut(old_commit_id);
+        if old_map.is_none() {
+            return;
+        }
+        let old_map = old_map.unwrap();
+        let removed_original_line_number = old_map.remove(&old_local_line_number);
+        if removed_original_line_number.is_none() {
+            return;
+        }
+        let removed_original_line_number = removed_original_line_number.unwrap();
+        if self.local_line_map.contains_key(new_commit_id) {
+            self.local_line_map
+                .get_mut(new_commit_id)
+                .unwrap()
+                .insert(new_local_line_number, removed_original_line_number);
+        } else {
+            let mut new_map = HashMap::new();
+            new_map.insert(new_local_line_number, removed_original_line_number);
+            self.local_line_map.insert(new_commit_id.clone(), new_map);
+        }
+    }
+
+    /// Once we've looked at all parents of a commit, any leftover lines must be
+    /// original to the current commit, so we save this information in
+    /// original_line_map.
+    fn drain_remaining_for_commit_id(&mut self, commit_id: &CommitId) {
+        self.file_cache.remove(commit_id);
+        let remaining_lines = self.local_line_map.remove(commit_id);
+        if remaining_lines.is_none() {
+            return;
+        }
+        let remaining_lines = remaining_lines.unwrap();
+        for (_, original_line_number) in remaining_lines {
+            self.original_line_map
+                .insert(original_line_number, commit_id.clone());
+        }
+    }
+
+    fn convert_to_results(self, original_contents: &[u8]) -> AnnotateResults {
+        let original_content_lines: Vec<String> =
+            original_contents.lines().map(|s| s.unwrap()).collect();
+        let mut result_lines = Vec::new();
+        for (idx, line) in original_content_lines.into_iter().enumerate() {
+            result_lines.push((self.original_line_map.get(&idx).unwrap().clone(), line));
+        }
+        AnnotateResults {
+            file_annotations: result_lines,
+        }
+    }
+
+    /// loads a given file into the cache under a specific commit id.
+    /// If there is already a file for a given commit, it is a no-op.
+    fn load_file_into_cache(
+        &mut self,
+        store: &Store,
+        commit_id: &CommitId,
+        file_path: &RepoPath,
+        file_id: &FileId,
+    ) -> BackendResult<()> {
+        if self.file_cache.contains_key(commit_id) {
+            return Ok(());
+        }
+
+        let file_contents = get_file_contents(store, file_path, file_id)?;
+        self.file_cache.insert(commit_id.clone(), file_contents);
+
+        Ok(())
+    }
+}
+
+/// Get line by line annotations for a specific file path in the repo.
+pub fn get_annotation_for_file(
+    file_path: &RepoPath,
+    repo: &ReadonlyRepo,
+    starting_commit: &Commit,
```

Comment:
nit: `repo` can be `&dyn Repo`. Annotation should work even if the repo is a MutableRepo.

And I would reorder arguments as `repo, starting_commit, file_path` (global context first.)